appsrc/dcfile/dcdirdmp.man
.TH DCDIRDMP 1 "1 May 2011" "DICOM PS3" "DICOM PS3 - Describe DICOMDIR content"
.SH NAME
dcdirdmp \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Describe DICOMDIR content
.SH SYNOPSIS
.HP 10
.B dcdirdmp
[
.B \-v|verbose|vv|veryverbose
]
[
.B \-showrecordinfo
]
[
.B \-description
]
[
.B \-p|paths
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dcdirdmp
reads the named dicom input file and interprets it as a DICOMDIR file, that
is a DICOM file containing an instance of the Basic Directory IOD.
.LP
If the file contains a valid instance of the directory, the hierarchical
record structure is traversed and displayed in a depth first manner. The
default output displays selected attribute values for recognized types
of directory record, including the Referenced File ID. The verbose options
increase the amount of information dumped.
.LP
Note that directory record offset values within attributes are in bytes
from the start of the file (including meta-information header and
pre-amble) and these point to the beginning of the Item delimiter that
introduces each item in the directory record sequence. The offset values
displayed in the verbose output for the first attribute of each sequence
item will therefore be 8 bytes greater, as the Item delimiter itself
is not displayed.
.SH OPTIONS
The description and verbose output go to standard error.
.LP
Binary attributes are written in hexadecimal with a preceding
"0x". Numeric string attributes are written in decimal. Attribute values
are displayed in hexadecimal or string format as determined by the value representation.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-v|verbose
.RS
Display all attributes in each directory record.
.RE
.TP
.B \-vv|veryverbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading and once read.
.RE
.TP
.B \-showrecordinfo
.RS
Display byte offset with each directory record entry, in hexadecimal.
.RE
.TP
.B \-description
.RS
Display the Study and Series Description attribute values.
.RE
.TP
.B \-p|paths
.RS
Show Unix style path information for each referenced file only (suppresses normal output).
Used, for example, to feed paths as command line arguments to some other utility.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcdirdmp DICOMDIR
.RE
PATIENT SMITH^^^^ 1634-94
.RE
        STUDY 1  19960101 113013 
.RE
                VISIT
.RE
                STUDY COMPONENT XA ?
.RE
                SERIES 1 XA
.RE
                        IMAGE 2
.RE
                         -> XA000002
.RE

.RE
% dcdirdmp -description DICOMDIR
.RE
PATIENT Jones^Thomas 3648263
.RE
        STUDY 1  20110101 103425 CT Chest/Abdo/Pelvis 
.RE
                SERIES 1 CT Portal Venous Phase
.RE
                        IMAGE 1
.RE
                         -> DICOM\\IM1
.RE

.RE
% dcdirdmp -v DICOMDIR
.RE
PATIENT SMITH^^^^ 1634-94
.RE
(0x0004,0x1400) UL Next Directory Record Offset ...
.RE
 ...
.RE
(0x0010,0x0010) PN Patient's Name VR=<PN> ... 
.RE
 ...
.RE
(0x0010,0x0020) LO Patient's ID ... 
.RE
(0x0010,0x0030) DA Patient's Birth Date ...
.RE
(0x0010,0x0040) CS Patient's Sex  ...
.RE
        STUDY 1  19960101 113013 
.RE
        (0x0004,0x1400) UL Next Directory Record Offset ...
.RE
 ...
.RE
        (0x0008,0x0050) SH Accession Number ...
.RE
...
.RE
\ 
.RE
% dcdirdmp -vv dicomdir.
.RE
******** While reading ... ********
.RE
@0x00000084,0x00000000 of 0xffffffff: (0x0002,0x0000) ...
.RE
 ...
.RE
RootDirectoryFirstRecord = 0x190
.RE
RootDirectoryLastRecord = 0x190
.RE
Number of records = 7
.RE
Offset = 0x190
.RE
        @0x00000198: (0x0004,0x1400) UL Next Dir Record ... 
.RE
...
.RE
Offset = 0x2a4
.RE
        @0x000002ac: (0x0004,0x1400) UL Next Dir Record ...
.RE
 ...
.RE
PATIENT 1634-94-1 1634-94
.RE
@0x00000198: (0x0004,0x1400) UL Next Dir Record Offset ...
.RE
 ...
.RE
@0x00000268: (0x0010,0x0010) PN Patient's Name ...
.RE
 ...
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcdump(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2011. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
.LP
MRDR not yet handled properly.


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcdirdmp.cc
#include "attrmxls.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "dicomdir.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool veryverbose=options.get("veryverbose") || options.get("vv");
	bool verbose=options.get("verbose") || options.get("v") || veryverbose;
	bool showrecordinfo=options.get("showrecordinfo");
	bool showdescription=options.get("description");
	bool showpaths=options.get("paths") || options.get("p");
	bool showabstract = !showpaths;
	bool showfilename=options.get("filename");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-v|-verbose|-vv|-veryverbose]"
			<< " [-showrecordinfo]"
			<< " [-description]"
			<< " [-p|-paths]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (veryverbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,veryverbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	//if (veryverbose) log << "******** As read ... ********" << endl; 
	//log << list;
	//list.write(log,veryverbose);

	success=success&&parseDicomdir(list,log,verbose,veryverbose,showrecordinfo,showabstract,showpaths,showdescription);

	return success ? 0 : 1;
}

